oak_crypto/src/encryptor.rs

```diff
@@ -0,0 +1,264 @@
+
+use crate::{
+    hpke::{
+        setup_base_recipient, setup_base_sender, KeyPair, RecipientRequestContext,
+        RecipientResponseContext, SenderRequestContext, SenderResponseContext,
+    },
+    schema::{AeadEncryptedMessage, EncryptedRequest, EncryptedResponse},
+};
+use alloc::{boxed::Box, sync::Arc, vec::Vec};
+use anyhow::{anyhow, Context};
+
+// TODO(#3642): Remove this key and use a real encapsulated key generated by the client once Java
+// encryption is implemented.
+pub(crate) const TEST_ENCAPSULATED_PUBLIC_KEY: [u8; 65] = [
+    4, 61, 141, 127, 160, 162, 184, 158, 72, 237, 105, 64, 182, 118, 163, 183, 174, 1, 81, 66, 139,
+    37, 218, 208, 17, 139, 159, 158, 68, 123, 124, 96, 114, 150, 38, 251, 112, 28, 121, 132, 45,
+    250, 118, 208, 142, 153, 124, 192, 139, 178, 239, 188, 177, 219, 52, 178, 123, 117, 254, 171,
+    248, 77, 4, 242, 118,
+];
+
+/// Info string used by Hybrid Public Key Encryption;
+pub(crate) const OAK_HPKE_INFO: &[u8] = b"Oak Hybrid Public Key Encryption v1";
+
+pub trait CryptoContextGenerator {
+    fn generate_context(&self) -> anyhow::Result<CryptoContext>;
+}
+
+// TODO(#3841): Implement Oak Kernel Crypto API and return corresponding session keys instead.
+pub struct CryptoContext {
+    pub(crate) recipient_request_context: RecipientRequestContext,
+    pub(crate) recipient_response_context: RecipientResponseContext,
+}
+
+pub struct EncryptionKeyProvider {
+    key_pair: KeyPair,
+}
+
+impl Default for EncryptionKeyProvider {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl EncryptionKeyProvider {
+    /// Creates a crypto provider with a newly generated key pair.
+    pub fn new() -> Self {
+        Self {
+            key_pair: KeyPair::generate(),
+        }
+    }
+
+    /// Returns a NIST P-256 SEC1 encoded point public key.
+    /// <https://secg.org/sec1-v2.pdf>
+    pub fn get_serialized_public_key(&self) -> Vec<u8> {
+        self.key_pair.get_serialized_public_key()
+    }
+}
+
+impl CryptoContextGenerator for EncryptionKeyProvider {
+    fn generate_context(&self) -> anyhow::Result<CryptoContext> {
+        let (recipient_request_context, recipient_response_context) =
+            setup_base_recipient(&TEST_ENCAPSULATED_PUBLIC_KEY, &self.key_pair, OAK_HPKE_INFO)
+                .context("couldn't generate recipient crypto context")?;
+        Ok(CryptoContext {
+            recipient_request_context,
+            recipient_response_context,
+        })
+    }
+}
+
+/// Encryptor object for encrypting sender requests that will be sent to the recipient and
+/// decrypting recipient responses that are received by the sender. Each Encryptor object
+/// corresponds to a single crypto session between the sender and the recipient.
+///
+/// Sequence numbers for requests and responses are incremented separately, meaning that there could
+/// be multiple responses per request and multiple requests per response.
+pub struct SenderEncryptor {
```

Comment:
Sender and Recipient are the terms from the HPKE RFC: https://www.rfc-editor.org/rfc/rfc9180.html#name-notation

Reply:
What do you think are the better names for these?
- `ClientEncryptor` and `ServerEncryptor`
- `ClientEncryptor` and `EnclaveEncryptor` (though `ClientEncryptor` will also be used in an enclave for enclave-to-enclave communication)
- `ClientEncryptor` and `ServiceEncryptor`

Reply:
FWIW I lean towards client / server.

Reply:
I'll leave `Sender` and `Recipient` for internal `hpke.rs` (so it'll be closer to the RFC which for example has a `SetupBase<ROLE>` function where "The ROLE template parameter is either S or R, depending on the role of sender or recipient, respectively").

But will also rename external API to `ClientEncryptor` and `ServerEncryptor`, so that it's more understandable, since these are the classes that will be used in Oak.

Reply:
Sounds good.

Reply:
Done